pkg/controller/workload/workload_processor_test.go

```diff
@@ -181,10 +181,35 @@ func Test_handleServiceWithWaypoint(t *testing.T) {
 	// Waypoint with network address.
 	svc1 := createFakeService("svc1", "10.240.10.1", "10.240.10.200", createLoadBalancing(workloadapi.LoadBalancing_UNSPECIFIED_MODE, make([]workloadapi.LoadBalancing_Scope, 0)))
 	// Waypoint with hostname.
-	svc2 := createFakeService("svc2", "10.240.10.1", "gateway.example.com/default", createLoadBalancing(workloadapi.LoadBalancing_UNSPECIFIED_MODE, make([]workloadapi.LoadBalancing_Scope, 0)))
+	svc2 := createFakeService("svc2", "10.240.10.2", "default/waypoint.default.svc.cluster.local", createLoadBalancing(workloadapi.LoadBalancing_UNSPECIFIED_MODE, make([]workloadapi.LoadBalancing_Scope, 0)))
 
 	assert.NoError(t, p.handleService(svc1))
 	assert.NoError(t, p.handleService(svc2))
+
+	// Front end map includes svc1 but not svc2 as its waypoint is not resolved.
+	svc1ID := checkFrontEndMap(t, svc1.Addresses[0].Address, p)
+	checkServiceMap(t, p, svc1ID, svc1, 0, 0)
+
+	checkNotExistInFrontEndMap(t, svc2.Addresses[0].Address, p)
+
+	waypointIP := "10.240.10.3"
+	waypointsvc := createFakeService("waypoint", waypointIP, "", createLoadBalancing(workloadapi.LoadBalancing_UNSPECIFIED_MODE, make([]workloadapi.LoadBalancing_Scope, 0)))
+	assert.NoError(t, p.handleService(waypointsvc))
+	updateWaypointOfService(svc2, waypointIP)
+
+	// Front end map includes svc2 and waypointsvc now.
+	svc2ID := checkFrontEndMap(t, svc2.Addresses[0].Address, p)
+	checkServiceMap(t, p, svc2ID, svc2, 0, 0)
+	wID := checkFrontEndMap(t, waypointsvc.Addresses[0].Address, p)
+	checkServiceMap(t, p, wID, waypointsvc, 0, 0)
+
+	// Insert svc whose waypoint hostname can be resolved directly.
+	svc3 := createFakeService("svc3", "10.240.10.4", "default/waypoint.default.svc.cluster.local", createLoadBalancing(workloadapi.LoadBalancing_UNSPECIFIED_MODE, make([]workloadapi.LoadBalancing_Scope, 0)))
+	assert.NoError(t, p.handleService(svc3))
+
+	updateWaypointOfService(svc3, waypointIP)
```

Comment:
don't you feel the test cases is not friendly to add, it has tight dependency with previous cases?

Reply:
For other cases that are not related to testing waypoint hostname resolution, we could create a a fake service with IP address type, just like before, then we don't have to call `updateWaypointOfService` everytime.